File: app/src/main/java/me/imzack/app/ender/view/adapter/TypeSearchListAdapter.java
package me.imzack.app.ender.view.adapter;

import android.graphics.Color;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import me.imzack.app.ender.App;
import me.imzack.app.ender.R;
import me.imzack.app.ender.model.DataManager;
import me.imzack.app.ender.model.bean.Type;
import me.imzack.app.ender.util.ResourceUtil;
import me.imzack.app.ender.util.StringUtil;
import me.imzack.app.ender.view.widget.CircleColorView;

import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;

public class TypeSearchListAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {

    private static final int TYPE_HEADER = 0;
    private static final int TYPE_ITEM = 1;

    private DataManager mDataManager;
    private List<Type> mTypeSearchList;
    private OnTypeItemClickListener mOnTypeItemClickListener;

    public TypeSearchListAdapter(DataManager dataManager, List<Type> typeSearchList) {
        mDataManager = dataManager;
        mTypeSearchList = typeSearchList;
    }

    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        switch (viewType) {
            case TYPE_HEADER:
                return new HeaderViewHolder(LayoutInflater.from(parent.getContext()).inflate(R.layout.header_list_type_search, parent, false));
            case TYPE_ITEM:
                return new ItemViewHolder(LayoutInflater.from(parent.getContext()).inflate(R.layout.item_list_type_search, parent, false));
            default:
                throw new IllegalArgumentException("The argument viewType cannot be " + viewType);
        }
    }

    @Override
    public void onBindViewHolder(RecyclerView.ViewHolder holder, int position) {
        switch (getItemViewType(position)) {
            case TYPE_HEADER:
                HeaderViewHolder headerViewHolder = (HeaderViewHolder) holder;
                headerViewHolder.mTypeSearchCountText.setText(ResourceUtil.getQuantityString(R.string.text_type_search, R.plurals.text_type_count, mTypeSearchList.size()));
                break;
            case TYPE_ITEM:
                final ItemViewHolder itemViewHolder = (ItemViewHolder) holder;
                Type type = mTypeSearchList.get(position - 1);
                itemViewHolder.mTypeMarkIcon.setFillColor(Color.parseColor(type.getTypeMarkColor()));
                itemViewHolder.mTypeMarkIcon.setInnerIcon(type.getTypeMarkPattern() == null ? null : App.getContext().getDrawable(ResourceUtil.getDrawableResourceId(type.getTypeMarkPattern())));
                itemViewHolder.mTypeMarkIcon.setInnerText(StringUtil.getFirstChar(type.getTypeName()));
                itemViewHolder.mTypeNameText.setText(type.getTypeName());
                if (mOnTypeItemClickListener != null) {
                    itemViewHolder.itemView.setOnClickListener(new View.OnClickListener() {
                        @Override
                        public void onClick(View v) {
                            mOnTypeItemClickListener.onTypeItemClick(mDataManager.getTypeLocationInTypeList(mTypeSearchList.get(itemViewHolder.getLayoutPosition() - 1).getTypeCode()));
                        }
                    });
                }
                break;
        }
    }

    @Override
    public int getItemCount() {
        return mTypeSearchList.size() + 1;
    }

    @Override
    public int getItemViewType(int position) {
        return position == 0 ? TYPE_HEADER : TYPE_ITEM;
    }

    class HeaderViewHolder extends RecyclerView.ViewHolder {
        @BindView(R.id.text_type_search_count)
        TextView mTypeSearchCountText;

        public HeaderViewHolder(View itemView) {
            super(itemView);
            ButterKnife.bind(this, itemView);
        }
    }

    class ItemViewHolder extends RecyclerView.ViewHolder {
        @BindView(R.id.ic_type_mark)
        CircleColorView mTypeMarkIcon;
        @BindView(R.id.text_type_name)
        TextView mTypeNameText;

        public ItemViewHolder(View itemView) {
            super(itemView);
            ButterKnife.bind(this, itemView);
        }
    }

    public interface OnTypeItemClickListener {
        void onTypeItemClick(int typeListPos);
    }

    public void setOnTypeItemClickListener(OnTypeItemClickListener listener) {
        mOnTypeItemClickListener = listener;
    }
}

File: app/src/main/kotlin/me/imzack/app/end/view/adapter/TypeSearchListAdapter.kt
package me.imzack.app.end.view.adapter

import android.graphics.Color
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import kotlinx.android.extensions.LayoutContainer
import kotlinx.android.synthetic.main.header_list_type_search.*
import kotlinx.android.synthetic.main.item_list_type_search.*
import me.imzack.app.end.R
import me.imzack.app.end.common.Constant
import me.imzack.app.end.model.DataManager
import me.imzack.app.end.model.bean.Type
import me.imzack.app.end.util.ResourceUtil
import me.imzack.app.end.util.StringUtil

class TypeSearchListAdapter(private val mTypeSearchList: List<Type>) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {
    
    var mOnTypeItemClickListener: ((typeListPos: Int) -> Unit)? = null

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder =
            when (viewType) {
                Constant.VIEW_TYPE_HEADER -> HeaderViewHolder(LayoutInflater.from(parent.context).inflate(R.layout.header_list_type_search, parent, false))
                Constant.VIEW_TYPE_ITEM -> ItemViewHolder(LayoutInflater.from(parent.context).inflate(R.layout.item_list_type_search, parent, false))
                else -> throw IllegalArgumentException("The argument \"viewType\" cannot be \"$viewType\"")
            }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        when (getItemViewType(position)) {
            Constant.VIEW_TYPE_HEADER -> {
                val headerViewHolder = holder as HeaderViewHolder
                headerViewHolder.text_type_search_count.text = ResourceUtil.getQuantityString(R.string.text_type_search, R.plurals.text_type_count, mTypeSearchList.size)
            }
            Constant.VIEW_TYPE_ITEM -> {
                val itemViewHolder = holder as ItemViewHolder
                val (_, name, markColor, markPattern, _, hasMarkPattern) = mTypeSearchList[position - 1]
                itemViewHolder.ic_type_mark.setFillColor(Color.parseColor(markColor))
                itemViewHolder.ic_type_mark.setInnerIcon(if (hasMarkPattern) ResourceUtil.getDrawable(markPattern!!) else null)
                itemViewHolder.ic_type_mark.setInnerText(StringUtil.getFirstChar(name))
                itemViewHolder.text_type_name.text = name
                itemViewHolder.itemView.setOnClickListener { mOnTypeItemClickListener?.invoke(DataManager.getTypeLocationInTypeList(mTypeSearchList[itemViewHolder.layoutPosition - 1].code)) }
            }
        }
    }

    override fun getItemCount() = mTypeSearchList.size + 1

    override fun getItemViewType(position: Int) = if (position == 0) Constant.VIEW_TYPE_HEADER else Constant.VIEW_TYPE_ITEM

    class HeaderViewHolder(override val containerView: View) : RecyclerView.ViewHolder(containerView), LayoutContainer

    class ItemViewHolder(override val containerView: View) : RecyclerView.ViewHolder(containerView), LayoutContainer
}
